site/content/blog/2020-03-15-instrumenting-go-code-via-ast-2.md

```diff
@@ -0,0 +1,182 @@
+---
+title: "Instrumenting Go code via AST, Part 2"
+slug: instrumenting-go-code-via-ast-2
+series: "AST"
+date: 2020-03-15T00:00:00-04:00
+author: Eli Yukelzon
+github: reflog
+community: eli.yukelzon
+---
+
+## Welcome!
+
+This is a second part of our AST blog post series, expanding on the subject of utilizing Go AST libraries libraries to automate and improve your workflow.
+
+In this post I'll discuss a rather common problem that comes up while working with Go code and the way we've solved it by sprinkling a little bit of AST magic dust. Let's dive in.
+
+## Problem: A struct with no interface
+
+Let's say you are working on a large code base that was not built with interfaces in mind, meaning, there are `structs` and methods attached to that `struct`, but there is not `interface` describing it. This is a perfectly valid approach when no you don't need to mock/stub the method implementations provided by that struct, or when there isn't more than one implementation of the same 'contract'. When these things **are** needed - we need to provide an `interface`.
+
+Here's a small code snippet to demonstrate:
+
+```go
+
+type Person struct {
+	Name string
+}
+
+func (p Person) Hello() string {
+	return "Hello: " + p.Name
+}
+
+type Animal struct {
+	Legs int
+}
+
+func (a Animal) Hello() string {
+	return fmt.Sprintf("I have %d legs!", a.Legs)
+}
+
+func main() {
+	p := Person{Name: "Fred"}
+	a := Animal{Legs: 4}
+	// ...	
+}
+```
+
+In this example both `Person` and `Animal` have the same `Hello()` method. If we wanted to store a list of **both** `Person` and `Animal` structs, we would have to define it as:
+```go
+list := []interface{}{p,a}
+```
+But this way we loose the type information of the list elements.
+This is where interfaces come in. Since both `Person` and `Animal` _implement_ a method with the same signature, we can extract that signature into an interface and use it for storing items in a list:
+```go
+type interface Hello {
+	Hello() string
+}
+// ...
+list := []Hello{p,a}
+fmt.Printf("Person: [%v] Animal: [%v]\n", list[0].Hello(), list[1].Hello())
+```
+
+Awesome. Now let's say the struct you are extracting the interface from is a big on. A really big one. With lots and lots of methods spread out in different `.go` files. Creating such interface manually would be very laborious.  
```

Comment:
Why does the interface have to contain _all_ methods of a struct? In general the user of the interface only cares about the methods that are needed. Interfaces are there to allow abstraction, but if every public method is on the interface, there is not abstraction.